mssmt/compacted_tree.go

```diff
@@ -392,3 +392,166 @@ func (t *CompactedTree) MerkleProof(ctx context.Context, key [hashSize]byte) (
 
 	return NewProof(proof), nil
 }
+
+// collectLeaves recursively traverses the compacted tree represented by the
+// given node and collects all non-empty leaf nodes along with their keys.
+func collectLeaves(ctx context.Context, tx TreeStoreViewTx, node Node) (
+	map[[hashSize]byte]*LeafNode, error) {
+
+	// Base case: If it's a compacted leaf node.
+	if compactedLeaf, ok := node.(*CompactedLeafNode); ok {
+		// We only care about non-empty leaves.
+		if compactedLeaf.LeafNode.IsEmpty() {
+			return make(map[[hashSize]byte]*LeafNode), nil
+		}
+
+		return map[[hashSize]byte]*LeafNode{
+			compactedLeaf.Key(): compactedLeaf.LeafNode,
+		}, nil
+	}
+
+	// Recursive step: If it's a branch node.
+	if _, ok := node.(*BranchNode); ok {
+		return collectLeavesRecursive(ctx, tx, node, 0)
+	}
+
+	// Handle unexpected node types (like ComputedNode if store returns
+	// them). If it's an empty leaf node implicitly (e.g., EmptyTree node),
+	// return empty.
+	if IsEqualNode(node, EmptyLeafNode) {
+		return make(map[[hashSize]byte]*LeafNode), nil
+	}
+
+	// Check against EmptyTree branches requires depth. If we encounter an
+	// unexpected node that isn't identifiable as empty, return error.
+	return nil, fmt.Errorf("unexpected node type %T encountered "+
+		"during leaf collection", node)
+}
+
+// collectLeavesRecursive is the helper for collectLeaves that includes depth.
+func collectLeavesRecursive(ctx context.Context, tx TreeStoreViewTx, node Node,
```

Comment:
I think this function used to be a helper for `collectLeaves`, but now it is the go-to. So I think doc is stale.